### PR TITLE
feat: store chat images at cloud store 

### PR DESCRIPTION
**Pull request by evgeniy-ac**
#177

### CODE REVIEW

**Comment by evgeniy-ac** [comment]
image

File: src/dotnet/Core/Blobs/ContentSaver.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+    public ContentSaver(IBlobStorageProvider blobStorageProvider)
+        => _blobStorage = blobStorageProvider.GetBlobStorage(BlobScope.ContentRecord);
+
+    public async Task SaveContent(Content content, CancellationToken token)
+    {
+        await _blobStorage.WriteAsync(content.ContentId, content.Stream, false, token).ConfigureAwait(false);
+        var blob = (await _blobStorage.GetBlobsAsync(new[] { content.ContentId }, token).ConfigureAwait(false)).Single();
+        if (blob == null)
+            throw new InvalidOperationException("Application invariant violated");
```
Comment (alexis-kochetov) [inline src/dotnet/Core/Blobs/ContentSaver.cs:17]: Let's provide more specific exception there

**Review by alexis-kochetov — APPROVED** [review]

File: src/dotnet/Core/Blobs/ContentSaver.cs (hunk 1 above)
Comment (evgeniy-ac) [reply src/dotnet/Core/Blobs/ContentSaver.cs]: done

File: src/dotnet/UI.Blazor/Components/ChatIcon/chat-icon.css [hunk 2]
```diff
@@ -1,7 +1,35 @@
...
+.chat-icon-size-20 {
+    @apply w-20 h-20 text-2xl;
+}
+
+.chat-icon-size-20 .chat-icon-picker {
+    @apply w-9 h-9;
+}
+
+.chat-icon-input-file-label {
+    @apply w-full h-full cursor-pointer flex items-center justify-center;
```
Comment (alexyakunin) [inline src/dotnet/UI.Blazor/Components/ChatIcon/chat-icon.css:34]: Check out how to sort CSS classes in ui-style channel

**Review by alexyakunin — COMMENTED** [review]
LGTM with minor comments

File: src/dotnet/Chat.Service/Controllers/ChatPicturesController.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            return BadRequest("Too many files");
+
+        var file = httpRequest.Form.Files[0];
+        if (file.Length == 0)
+            return BadRequest("Image is empty");
+
+        if (file.Length > Constants.Chat.PictureFileSizeLimit)
+            return BadRequest("Image is too big");
+
+        // TODO Does not work
```
[the remark this answers is not in the document]
Comment (evgeniy-ac) [reply src/dotnet/Chat.Service/Controllers/ChatPicturesController.cs:39]: fixed

File: src/dotnet/Chat.Service/Controllers/ChatPicturesController.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        IChats chats,
+        IContentSaver contentSaver)
+    {
+        _sessionResolver = sessionResolver;
+        _chats = chats;
+        _contentSaver = contentSaver;
+    }
+
+    [HttpPost, Route("api/chats/{chatId}/upload-picture")]
+    public async Task<IActionResult> UploadPicture(string chatId, CancellationToken token)
```
[the remark this answers is not in the document]
Comment (evgeniy-ac) [reply src/dotnet/Chat.Service/Controllers/ChatPicturesController.cs:23]: done

File: src/dotnet/Chat.Service/Controllers/ContentController.cs [hunk 5]
```diff
@@ -20,7 +20,7 @@ public async Task<ActionResult> Download(string blobId, CancellationToken cancel
             return NotFound();
         var blob = (await blobStorage.GetBlobsAsync(new[] {blobId}, cancellationToken).ConfigureAwait(false)).Single();
```
[the remark this answers is not in the document]
Comment (evgeniy-ac) [reply src/dotnet/Chat.Service/Controllers/ContentController.cs:21]: fixed

File: src/dotnet/Chat.UI.Blazor/Components/ChatRightPanel/ChatBasicSettings.razor [hunk 6]
```diff
@@ -67,4 +78,29 @@
             IsPublicId = $"{FormId}-is-public";
         }
     }
+
+    private async Task OnImagePicked(InputFileChangeEventArgs e) {
+        if (e.File.Size > Constants.Chat.PictureFileSizeLimit) {
+            ErrorUI.ShowError("File is too big. Max file size: 1Mb.");
```
[the remark this answers is not in the document]
Comment (evgeniy-ac) [reply src/dotnet/Chat.UI.Blazor/Components/ChatRightPanel/ChatBasicSettings.razor:84]: Done

File: src/dotnet/Core/Blobs/IContentSaver.cs [hunk 7]
```diff
@@ -0,0 +1,6 @@
+namespace ActualChat.Blobs;
+
+public interface IContentSaver
+{
+    Task SaveContent(Content content, CancellationToken token);
```
[the remark this answers is not in the document]
Comment (evgeniy-ac) [reply src/dotnet/Core/Blobs/IContentSaver.cs:5]: Done

File: src/dotnet/Core/Constants.cs [hunk 8]
```diff
@@ -8,11 +8,17 @@ public static class Chat
         public static TileStack<long> IdTileStack { get; } = TileStacks.Long16To1K;
         public static TileStack<Moment> TimeTileStack { get; } = TileStacks.Moment3MTo6Y;
         public static TimeSpan MaxEntryDuration { get; } = TimeTileStack.MinTileSize.EpochOffset; // 3 minutes, though it can be any
+        public const int PictureFileSizeLimit = 1 * 1024 * 1024;
     }
 
     public static class Attachments
     {
         public const int FileSizeLimit = 8 * 1024 * 1024;
         public const int FilesNumberLimit = 10;
```
[the remark this answers is not in the document]
Comment (evgeniy-ac) [reply src/dotnet/Core/Constants.cs:17]: Done

File: src/dotnet/UI.Blazor/Components/ChatIcon/ChatIcon.razor [hunk 9]
```diff
@@ -1,39 +1,76 @@
...
         var h = hash % 360;
         return $"hsl({h}, 30%, 80%)";
     }
+
+    private async Task OnPictureClick(MouseEventArgs arg) {
+        await PictureClick.InvokeAsync(arg);
+    }
+
+    private async Task OnFile(InputFileChangeEventArgs arg) {
+        await ImagePicked.InvokeAsync(arg);
```
[the remark this answers is not in the document]
Comment (evgeniy-ac) [reply src/dotnet/UI.Blazor/Components/ChatIcon/ChatIcon.razor:74]: Fixed

File: src/dotnet/UI.Blazor/Components/ChatIcon/chat-icon.css (hunk 2 above)
Comment (evgeniy-ac) [reply src/dotnet/UI.Blazor/Components/ChatIcon/chat-icon.css:34]: Done

File: src/dotnet/UI.Blazor/Components/ChatIcon/ChatIcon.razor [hunk 10]
```diff
@@ -1,39 +1,76 @@
...
     }
 
     private string GenerateColor() {
         var hash = Title.GetDjb2HashCode();
         var h = hash % 360;
         return $"hsl({h}, 30%, 80%)";
     }
+
+    private async Task OnPictureClick(MouseEventArgs arg) {
+        await PictureClick.InvokeAsync(arg);
```
[the remark this answers is not in the document]
Comment (evgeniy-ac) [reply src/dotnet/UI.Blazor/Components/ChatIcon/ChatIcon.razor:70]: Fixed

File: src/dotnet/Chat.UI.Blazor/Components/ChatRightPanel/ChatBasicSettings.razor [hunk 11]
```diff
@@ -67,4 +78,29 @@
...
+        var fileContent = new StreamContent(e.File.OpenReadStream(Constants.Chat.PictureFileSizeLimit));
+        fileContent.Headers.ContentType = new MediaTypeHeaderValue(e.File.ContentType);
+        content.Add(fileContent, "file", e.File.Name);
+        var client = ClientFactory.CreateClient();
+        var url = UriMapper.ToAbsolute($"/api/chats/{Chat.Id}/upload-picture");
+        var response = await client.PostAsync(url, content);
+        if (response.IsSuccessStatusCode) {
+            _form.Picture = await response.Content.ReadAsStringAsync();
+        } else {
+            ErrorUI.ShowError(response.ReasonPhrase ?? "Something went wrong");
```
[the remark this answers is not in the document]
Comment (evgeniy-ac) [reply src/dotnet/Chat.UI.Blazor/Components/ChatRightPanel/ChatBasicSettings.razor:98]: Removed

**Pull request merged**